codegen/codegen.cabal

```diff
@@ -51,16 +51,16 @@ executable codegen-exe
     , optparse-applicative >= 0.14.3.0
   extra-libraries:     stdc++
 
-test-suite doctests
-  default-language: Haskell2010
-  type:            exitcode-stdio-1.0
-  hs-source-dirs:  test
-  main-is:         doctests.hs
-  ghc-options:     -Wall -threaded
-  build-depends:
-      base
-    , doctest
-    , megaparsec
+--test-suite doctests
```

Comment:
haskell.nix unfortunately doesn't have a good story for doctests :-\

Reply:
Oh, I should add that you probably want to leave this `doctests` component enabled, but explicitly disable it within haskell.nix.

You can probably do it with a module like the following:

```nix
{
  packages.codegen.components.tests.doctests.buildable = false;
}
```

(Or something like this.  I haven't actually looked at the haskell.nix docs to confirm this...)

Reply:
Hm, I'm all in favor of this, but I've got some trouble with your suggested solution. After adding this configuration item to the `modules` section of `cabalProject`, I'm getting
```
$ nix-build ./shared.nix -A defaultCpu.hsPkgs.codegen.components.tests.doctests
error: The option `packages.codegen.components.tests.doctests.buildable' has conflicting definitions, in `/nix/store/mxzz2cj5a8c32b39jjd0qkk9vj8kp5l9-source/modules/plan.nix' and `<unknown-file>'.
```

Reply:
am I changing this attribute?
https://github.com/input-output-hk/haskell.nix/blob/master/modules/package.nix#L145

Reply:
I thought about this a little more.

Ideally what we want is for `haskell.nix` to build the doctest test, but not actually run it.

It should be possible with a change like this:

```diff
diff --git a/shared.nix b/shared.nix
index 1d9f322f..7318ced0 100644
--- a/shared.nix
+++ b/shared.nix
@@ -56,6 +56,8 @@ let
             ];
             packages.libtorch-ffi.flags.cuda = USE_CUDA;
             packages.libtorch-ffi.flags.gcc = USE_GCC;
+
+            packages.codegen.components.tests.doctest.doCheck = false;
           })
         ];
       };
```

However, for whatever reason, this doesn't appear to actually work.  The `codegen` doctests appear to be run when you build the `checks`:

```console
$ nix-build shared.nix -A defaultCpu.hsPkgs.codegen.checks
trace: Using latest index state for hasktorch!
trace: Using index-state: 2020-06-11T00:00:00Z for hasktorch
trace: WARNING: license "GPL-2.0-or-later AND BSD-3-Clause" not found
these derivations will be built:
  /nix/store/2rhf2f9bd7c2rrn5i11wsdn58ffyzbvi-codegen-test-doctests-0.1.0.0-check.drv
building '/nix/store/2rhf2f9bd7c2rrn5i11wsdn58ffyzbvi-codegen-test-doctests-0.1.0.0-check.drv'...
unpacking sources
unpacking source archive /nix/store/7wf5xaqr82vv0wnyvl651cq9yb8gr562-codegen-test-doctests-0.1.0.0
source root is codegen-test-doctests-0.1.0.0
building

<no location info>: error: can't find file: src/ParseFunctionSig.hs

builder for '/nix/store/2rhf2f9bd7c2rrn5i11wsdn58ffyzbvi-codegen-test-doctests-0.1.0.0-check.drv' failed with exit code 1
error: build of '/nix/store/2rhf2f9bd7c2rrn5i11wsdn58ffyzbvi-codegen-test-doctests-0.1.0.0-check.drv' failed
```

However, I don't know why this would be happening, since the upstream `haskell.nix` seems pretty clear that if you set `doCheck` for the test components, the checks aren't actually run:

https://github.com/input-output-hk/haskell.nix/blob/643c219a5ef0a4b2570c8d056fa827214b4a0b64/builder/hspkg-builder.nix#L100-L102

You can see that checks with `doCheck` set to `false` should be getting filtered out of the `codegen.checks` attribute set, however that doesn't appear to be happening.

(Oh, and I'm not sure if this was clear, but keep in mind that when you build the `test` components, the tests don't actually get run.  You have to explicitly build the `check` attribute.  You can play around with this with a command like `nix-build shared.nix -A defaultCpu.hsPkgs.codegen.components.tests.spec` vs `nix-build shared.nix -A defaultCpu.hsPkgs.codegen.checks.spec`)
-----------------------------

So I'm not really sure what to suggest here. As long as you don't explicitly call `defaultCpu.hsPkgs.codegen.checks.spec` anywhere, I guess you won't have any problems.

Reply:
Ooops, looks like I made a mistake above.  I had written `packages.codegen.components.tests.doctest.doCheck`, but that is incorrect.

Here is the correct diff:

```diff
diff --git a/shared.nix b/shared.nix
index 1d9f322f..7318ced0 100644
--- a/shared.nix
+++ b/shared.nix
@@ -56,6 +56,8 @@ let
             ];
             packages.libtorch-ffi.flags.cuda = USE_CUDA;
             packages.libtorch-ffi.flags.gcc = USE_GCC;
+
+            packages.codegen.components.tests.doctests.doCheck = false;
           })
         ];
       };
```

It should be `doctests` instead of `doctest`.

With this set correctly, the `doctest` test won't actually get run:

```console
$ nix-build shared.nix -A defaultCpu.hsPkgs.codegen.checks
/nix/store/ga351x83q4kayhqadw6rrrpp4zzbx7gk-codegen-test-spec-0.1.0.0-check
```

----------------------------------------

Also, unrelated, but to answer your question above:

> ```console
> $ nix-build ./shared.nix -A defaultCpu.hsPkgs.codegen.components.tests.doctests
> error: The option `packages.codegen.components.tests.doctests.buildable' has conflicting definitions, in `/nix/store/mxzz2cj5a8c32b39jjd0qkk9vj8kp5l9-source/modules/plan.nix' and `<unknown-file>'.
> ```

When working with nix modules, sometimes you need to use functions like `mkForce` to get things working:

```diff
diff --git a/shared.nix b/shared.nix
index 1d9f322f..c4c343b0 100644
--- a/shared.nix
+++ b/shared.nix
@@ -47,8 +47,8 @@ let
         src = pkgs.haskell-nix.haskellLib.cleanGit { name = "hasktorch"; src = ./.; };
         compiler-nix-name = haskellCompiler;
         modules = [
-          ({ config, ... }: {
-            # packages.codegen.components.tests.doctests.buildable = false;
+          ({ config, lib, ... }: {
+            packages.codegen.components.tests.doctests.buildable = lib.mkForce false;
             packages.libtorch-ffi.configureFlags = [
               "--extra-lib-dirs=${pkgs.torch}/lib"
               "--extra-include-dirs=${pkgs.torch}/include"
```